### PR TITLE
Add SD for Amazon EC2 instances

### PR DESCRIPTION
**Pull request by mjibson**
This PR adds support for scraping EC2 instances that are members of ELBs. We found this to be a better solution than just scraping all of our EC2 instances because our ELBs have differing instances with different ports, TLS, and other configuration details. This PR allows us to monitor each ELB correctly. We have successfully used this PR to monitor an etcd cluster running in an ELB.

Godeps in a second commit to make reviewing the actually added code easier. Will squash before merging.

### CODE REVIEW

**Comment by mjibson** [comment]
Could someone clue me into what's going on with the build failure? I've been trying to figure it out for a few hours. `make build` works. Building the dockerimage works. I'm not sure why just `make` is breaking. I can't even get `make` to work on the master branch, so I'm not sure it's caused by this change.

**Comment by brian-brazil** [comment]
I'm not sure about having this as a SD, as it only covers EC2 services that are running behind an ELB so we'll also need an EC2 SD to cover all the use cases. Each SD will require ongoing maintenance, so I think for AWS that we should just have EC2 and if that doesn't cover a need there's the option of file sd.

You're also hardcoding to the public IP address, which will break where that doesn't exist in some VPC setups and in general users will be monitoring from inside their own network. The private IP should be the default.

**Comment by juliusv** [comment]
@mjibson As for the build failure, that's an issue with how we vendor dependencies - it hasn't bitten us so far because we haven't vendored any packages with "internal" packages as transitive dependencies yet. But the AWS-SDK does have that... @fabxc is working on a general fix for this, which should be in master soon.

**Comment by mjibson** [comment]
I'd really like to avoid using a file SD for this, so hopefully we can come to some resolution. I understand that there are situations where the ELB-based lookup isn't sufficient. I think we can do what we need by matching tags of EC2 instances to a regex. Maybe a configuration file like:

```
- job_name: some_group
  ec2_sd_configs:
    - region: us-east-1
      access: aosentuh
      secret: aoesnuth
      port: 9092
      tags:
        Name: Service-.*
- job_name: ec2_etcd
  scheme: https
  ec2_sd_configs:
    - region: us-east-1
      access: aosentuh
      secret: aoesnuth
      port: 2379
      tags:
        Name: etcd-.*
```

Where the `tags` map specifies keys that must be present in an EC2 instance with values matching the regex. This would provide enough filtering for our needs. These tags would also then be added to the target.

**Comment by brian-brazil** [comment]
@mjibson That's pretty much what I'm thinking. We'd expose all the things the ec2 api provides including tags and security group and have relabelling do any required filtering.

**Comment by mjibson** [comment]
I've updated the PR to fetch EC2 instances by matching tag regex. Questions I have:

1. Is the `Sources()` func correct? I don't understand what it's for.
2. Are the base labels reasonable? Currently it's the full tags for anything matched in the config. (Not the regex value, the actual value that passed the match.) Not sure if this is good, but seems best since relabeling can remove anything a user doesn't want. However, maybe using the regex string is better instead.

File: config/config.go [hunk 1]
```diff
@@ -664,6 +671,43 @@ func (c *KubernetesSDConfig) UnmarshalYAML(unmarshal func(interface{}) error) er
 	return checkOverflow(c.XXX, "kubernetes_sd_config")
 }
 
+// EC2SDConfig is the configuration for EC2 based service discovery.
+type EC2SDConfig struct {
+	Region          string                    `yaml:"region"`
+	Access          string                    `yaml:"access,omitempty"`
```
Comment (brian-brazil) [inline config/config.go]: The full "AccessKey" would be clearer

File: config/config.go [hunk 2]
```diff
@@ -664,6 +671,43 @@ func (c *KubernetesSDConfig) UnmarshalYAML(unmarshal func(interface{}) error) er
...
+	Secret          string                    `yaml:"secret,omitempty"`
```
Comment (brian-brazil) [inline config/config.go]: Similarly, and make sure it's covered by the regex above that hides secrets

File: config/config.go [hunk 3]
```diff
@@ -664,6 +671,43 @@ func (c *KubernetesSDConfig) UnmarshalYAML(unmarshal func(interface{}) error) er
...
+	RefreshInterval Duration                  `yaml:"refresh_interval,omitempty"`
+	Port            int                       `yaml:"port"`
+	Tags            map[string]string         `yaml:"tags,omitempty"`
```
Comment (brian-brazil) [inline config/config.go]: Every option below here should be done with relabelling, we need to dump the entire ec2 instance info into meta labels in a sane fashion.

File: config/config.go [hunk 4]
```diff
@@ -128,6 +128,11 @@ var (
 		RequestTimeout: Duration(10 * time.Second),
 		RetryInterval:  Duration(1 * time.Second),
 	}
+
+	// DefaultEC2SDConfig is the default EC2 SD configuration.
+	DefaultEC2SDConfig = EC2SDConfig{
+		RefreshInterval: Duration(30 * time.Second),
```
Comment (brian-brazil) [inline config/config.go]: Instances take at least a minute to startup, so I'd say 60s is fine

File: config/config.go [hunk 5]
```diff
@@ -664,6 +671,43 @@ func (c *KubernetesSDConfig) UnmarshalYAML(unmarshal func(interface{}) error) er
...
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *EC2SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultEC2SDConfig
+	type plain EC2SDConfig
+	err := unmarshal((*plain)(c))
+	if err != nil {
+		return err
+	}
+	if c.Port == 0 {
```
Comment (brian-brazil) [inline config/config.go]: Maybe default this to 80?

File: retrieval/discovery/ec2.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+		tags:      tags,
+		regexTags: conf.RegexTags,
+		done:      make(chan struct{}),
+		interval:  time.Duration(conf.RefreshInterval),
+		port:      conf.Port,
+		public:    conf.Public,
+		instanceFilter: []*ec2.Filter{
+			{
+				Name:   aws.String("instance-state-name"),
+				Values: []*string{aws.String("running")},
```
Comment (brian-brazil) [inline retrieval/discovery/ec2.go]: This shouldn't be hardcoded

File: config/config.go (hunk 5 above)
Comment (mjibson) [inline config/config.go]: I followed used the dns retrival as a template, which does exactly this. Happy to set the default to 80, though.

File: retrieval/discovery/ec2.go [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+				}
+				missing := false
+				labels := make(model.LabelSet)
+				for k, r := range ed.regexTags {
+					v, ok := itags[k]
+					if !ok || !r.MatchString(v) {
+						missing = true
+						break
+					}
+					labels[model.LabelName(k)] = model.LabelValue(v)
```
Comment (brian-brazil) [inline retrieval/discovery/ec2.go]: tags shouldn't automatically become labels, the user should have to configure that with relabelling.

File: retrieval/discovery/ec2.go (hunk 6 above)
Comment (mjibson) [inline retrieval/discovery/ec2.go]: Why? If an instance isn't running it doesn't have an IP address.

File: config/config.go (hunk 3 above)
Comment (mjibson) [inline config/config.go]: I'm not sure I understand. RegexTags is a cache of Tags, but converted to regexps. Public is a user option to use the public vs private IP addresses. I'm don't understand what either of those have to do with relabeling.

File: retrieval/discovery/ec2.go (hunk 6 above)
Comment (brian-brazil) [inline retrieval/discovery/ec2.go]: Are you sure that's true for the stopping, shutting-down, stopped in VPC and rebooting starts?

Comment (mjibson) [inline retrieval/discovery/ec2.go]: No, I'm not sure. I can remove this. Thought it was an easy performance win since it limited the instances being returned to useable ones. I think the tag-key filter is good enough for that, though.

File: config/config.go (hunk 5 above)
Comment (brian-brazil) [inline config/config.go]: Probably best to keep consistent then.

File: config/config.go (hunk 3 above)
Comment (brian-brazil) [inline config/config.go]: There are many different field that users may want to filter on or have as labels. Rather than adding each as configuration variables we should supply them all as labels, and let the user use relabelling to do what they want.

File: retrieval/discovery/ec2.go (hunk 6 above)
Comment (grobie) [inline retrieval/discovery/ec2.go]: I'd recommend to at least exclude instances in state `terminated`. AWS returns them via the API for 1-2 days. I can also confirm that `stopped` instances don't have IPs, and I'm 90% sure for stopping and shutting-down. Issuing a reboot doesn't change the instance state from running.

Comment (brian-brazil) [inline retrieval/discovery/ec2.go]: > I can also confirm that stopped instances don't have IPs

Was that in VPC? EBS VPC instances keep their private IP across stops and starts, so they may be reported.

Comment (grobie) [inline retrieval/discovery/ec2.go]: You're right, I found some instances inside a VPC which still have their private IP.

I just had to shut down some instances (VPC) and they didn't have a private IP anymore. Though, it shouldn't really matter. Excluding instances in `terminated` state would be useful for the amount of instances we have though.

Comment (brian-brazil) [inline retrieval/discovery/ec2.go]: Excluding terminated would make sense.

File: config/config.go (hunk 3 above)
Comment (mjibson) [inline config/config.go]: I still don't understand how this plays into the public/private IP question, since which IP to use as the target doesn't come into play after relabeling. Or does it and I just haven't seen how that works?

File: config/config.go [hunk 8]
```diff
@@ -664,6 +671,33 @@ func (c *KubernetesSDConfig) UnmarshalYAML(unmarshal func(interface{}) error) er
 	return checkOverflow(c.XXX, "kubernetes_sd_config")
 }
 
+// EC2SDConfig is the configuration for EC2 based service discovery.
+type EC2SDConfig struct {
+	Region          string             `yaml:"region"`
+	AccessKey       string             `yaml:"access_key,omitempty"`
+	SecretKey       string             `yaml:"secret_key,omitempty"`
```
Comment (mjibson) [inline config/config.go]: You requested making sure this was in the regex to hide stuff. I was not able to find that regex or any reference to hiding. Could you point me to what you're talking about so I can add it?

**Comment by mjibson** [comment]
Updated with many fixes. Will work on the tag improvements next week.

File: config/config.go (hunk 8 above)
Comment (juliusv) [inline config/config.go]: He was referring to this regex:

https://github.com/prometheus/prometheus/blob/ec04ba38a2672786f69455b735fd43e6d06afa90/config/config.go#L36

...which is used by the status page config printer to "blur" out credentials.

File: config/config.go (hunk 3 above)
Comment (brian-brazil) [inline config/config.go]: Relabelling can affect what IP to use for scraping via the __address__ label.

**Comment by mjibson** [comment]
Updated with hidden regex and fixes for public/private IP and tags, which can now all be configured via relabeling:

```
scrape_configs:
- job_name: ec2
  relabel_configs:
    - source_labels: ['__meta_ec2_public_ip']
      target_label: __address__
      regex: .*
      replacement: ${0}
    - source_labels: ['__meta_ec2_tag_aws_cloudformation_stack_id']
      target_label: ec2_stack
      regex: .*
      replacement: ${0}
  ec2_sd_configs:
    - region: us-east-1
      access_key: <hidden>
      secret_key: <hidden>
      port: 2379
      tags:
        Name: foo-.*
        'aws:autoscaling:groupName': bar-.*
```

Will use the public IP (instead of the private, the default) and set the `ec2_stack` label.

File: retrieval/discovery/ec2.go [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+					addr := model.LabelValue(fmt.Sprintf("%s:%d", *inst.PrivateIpAddress, ed.port))
+					labels[ec2Label+"private_ip"] = addr
+					labels[model.AddressLabel] = addr
+				} else {
+					continue
+				}
+				itags := make(map[string]string)
+				for _, t := range inst.Tags {
+					itags[*t.Key] = *t.Value
+					name := cleanTag(*t.Key)
```
Comment (mjibson) [inline retrieval/discovery/ec2.go]: This was the least bad idea I had about how to fix up the EC2 tags to something valid for prometheus. Is there a better way?

Comment (brian-brazil) [inline retrieval/discovery/ec2.go]: That's pretty much what we usually do, you can use the existing SanitizeLabelName function for this.

File: retrieval/discovery/ec2.go [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+	}
+	if err := EC2.DescribeInstancesPages(&dii, func(p *ec2.DescribeInstancesOutput, lastPage bool) bool {
+		for _, r := range p.Reservations {
+			for _, inst := range r.Instances {
+				labels := model.LabelSet{
+					ec2Label + "instance_id": model.LabelValue(*inst.InstanceId),
+				}
+				if inst.PublicIpAddress != nil {
+					addr := model.LabelValue(fmt.Sprintf("%s:%d", *inst.PublicIpAddress, ed.port))
+					labels[ec2Label+"public_ip"] = addr
```
Comment (brian-brazil) [inline retrieval/discovery/ec2.go]: I wouldn't expect a label called "IP" to contain a port number

File: retrieval/discovery/ec2.go [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+				}
+				if inst.PrivateIpAddress != nil {
+					addr := model.LabelValue(fmt.Sprintf("%s:%d", *inst.PrivateIpAddress, ed.port))
+					labels[ec2Label+"private_ip"] = addr
+					labels[model.AddressLabel] = addr
+				} else {
+					continue
+				}
+				itags := make(map[string]string)
```
Comment (brian-brazil) [inline retrieval/discovery/ec2.go]: Avoid make where you can, so `map[string]string{}`

File: retrieval/discovery/ec2.go [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+				for _, t := range inst.Tags {
+					itags[*t.Key] = *t.Value
+					name := cleanTag(*t.Key)
+					labels[ec2Label+"tag_"+model.LabelName(name)] = model.LabelValue(*t.Value)
+				}
+				missing := false
+				for k, r := range ed.regexTags {
```
Comment (brian-brazil) [inline retrieval/discovery/ec2.go]: This can be deleted, relabelling will let you do this

File: retrieval/discovery/ec2.go [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+				if inst.PublicIpAddress != nil {
+					addr := model.LabelValue(fmt.Sprintf("%s:%d", *inst.PublicIpAddress, ed.port))
+					labels[ec2Label+"public_ip"] = addr
+				}
+				if inst.PrivateIpAddress != nil {
+					addr := model.LabelValue(fmt.Sprintf("%s:%d", *inst.PrivateIpAddress, ed.port))
+					labels[ec2Label+"private_ip"] = addr
+					labels[model.AddressLabel] = addr
+				} else {
+					continue
```
Comment (brian-brazil) [inline retrieval/discovery/ec2.go]: Can you add a comment on this?

File: retrieval/discovery/ec2.go (hunk 10 above)
Comment (mjibson) [inline retrieval/discovery/ec2.go]: Ok. So if a user wanted to use the public IP, they'd have to hard code the port into the replacement regex? I'm fine with that, just checking.

File: retrieval/discovery/ec2.go (hunk 13 above)
Comment (mjibson) [inline retrieval/discovery/ec2.go]: This one is tricky. If there's no private IP address and we leave the `model.AddressLabel` field blank, then the targets sent to prometheus will be `http://:80/`, which is useless. Hence, the continue. Is there a better way?

File: retrieval/discovery/ec2.go [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+			}
+		case <-done:
+			return
+		}
+	}
+}
+
+// Sources implements the TargetProvider interface.
+func (ed *EC2Discovery) Sources() []string {
+	return []string{*ed.aws.Region}
```
Comment (brian-brazil) [inline retrieval/discovery/ec2.go]: You need to return all the sources here too, or targets will blip.

@fabxc This is roughly the reason I'm in favour of a full dump from Sources() rather than just the source names.

Comment (mjibson) [inline retrieval/discovery/ec2.go]: I don't know what Sources are. Can you link me to some documentation?

File: retrieval/discovery/ec2.go (hunk 13 above)
Comment (mjibson) [inline retrieval/discovery/ec2.go]: If we keep this I'll invert the logic, though, so the continue is in the if and the other stuff in the containing block.

File: retrieval/discovery/ec2.go (hunk 10 above)
Comment (brian-brazil) [inline retrieval/discovery/ec2.go]: Yes, we only ask for a port in SD for providing the __address__ label as things will break otherwise.

File: config/config.go (hunk 5 above)
Comment (mjibson) [inline config/config.go]: I've changed my mind and now think it's best to default this to 80. This is because in the case where a user wants to use the public ip address via relabeling, the port is unused. Hence, I think we should make it optional (i.e., default to 80) so that it need not be specified when the public ip is used.

**Comment by mjibson** [comment]
Updated with simplified port and IP address handling. Removed regex tag filters since that can be done in relabeling.

File: config/config.go [hunk 15]
```diff
@@ -33,7 +33,7 @@ var (
 	patJobName    = regexp.MustCompile(`^[a-zA-Z_][a-zA-Z0-9_-]*$`)
 	patFileSDName = regexp.MustCompile(`^[^*]*(\*[^/]*)?\.(json|yml|yaml|JSON|YML|YAML)$`)
 	patRulePath   = regexp.MustCompile(`^[^*]*(\*[^/]*)?$`)
-	patAuthLine   = regexp.MustCompile(`((?:username|password|bearer_token):\s+)(".+"|'.+'|[^\s]+)`)
+	patAuthLine   = regexp.MustCompile(`((?:username|password|bearer_token|access_key|secret_key):\s+)(".+"|'.+'|[^\s]+)`)
```
Comment (brian-brazil) [inline config/config.go]: Username and access_key are not secrets

File: retrieval/discovery/ec2.go (hunk 14 above)
Comment (brian-brazil) [inline retrieval/discovery/ec2.go]: https://github.com/prometheus/prometheus/blob/master/retrieval/targetmanager.go#L36

File: config/config.go [hunk 16]
```diff
@@ -664,6 +671,32 @@ func (c *KubernetesSDConfig) UnmarshalYAML(unmarshal func(interface{}) error) er
...
+	SecretKey       string             `yaml:"secret_key,omitempty"`
+	RefreshInterval Duration           `yaml:"refresh_interval,omitempty"`
+	Port            int                `yaml:"port"`
+	Tags            map[string]*Regexp `yaml:"tags,omitempty"`
+	// Catches all undefined fields and must be empty after parsing.
+	XXX map[string]interface{} `yaml:",inline"`
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *EC2SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
```
Comment (brian-brazil) [inline config/config.go]: Check for region being present?

File: config/config.go (hunk 15 above)
Comment (mjibson) [inline config/config.go]: username was already present. Want me to remove them both?

File: retrieval/discovery/ec2.go (hunk 14 above)
Comment (mjibson) [inline retrieval/discovery/ec2.go]: Yes, I've read that but it's not helpful.

File: config/config.go (hunk 15 above)
Comment (brian-brazil) [inline config/config.go]: Yes please

File: retrieval/discovery/ec2.go (hunk 14 above)
Comment (mjibson) [inline retrieval/discovery/ec2.go]: Doesn't the existing code do exactly that? It sends the region as the Source here, then later on does the same during Run.

Comment (brian-brazil) [inline retrieval/discovery/ec2.go]: Ah I misread, usually this code is more complex.

Comment (mjibson) [inline retrieval/discovery/ec2.go]: Is it possible that multiple ec2_sd_configs that have the same region will mess with each other because of this?

File: config/config.go (hunk 16 above)
Comment (mjibson) [inline config/config.go]: Done, along with access and secret keys.

File: retrieval/discovery/ec2.go (hunk 14 above)
Comment (brian-brazil) [inline retrieval/discovery/ec2.go]: It's possible if there's multiple accounts within the same scrape config.

Comment (mjibson) [inline retrieval/discovery/ec2.go]: Recommended solution for this? Does each ec2_sd_config need to produce a unique Source? Or maybe the Source implementation here needs to add one source per EC2 instance, and use the instance-id as the source?

Comment (brian-brazil) [inline retrieval/discovery/ec2.go]: The instance-id as source would be the way to handle it.

File: config/config.go [hunk 17]
```diff
@@ -33,7 +33,7 @@ var (
 	patJobName    = regexp.MustCompile(`^[a-zA-Z_][a-zA-Z0-9_-]*$`)
 	patFileSDName = regexp.MustCompile(`^[^*]*(\*[^/]*)?\.(json|yml|yaml|JSON|YML|YAML)$`)
 	patRulePath   = regexp.MustCompile(`^[^*]*(\*[^/]*)?$`)
-	patAuthLine   = regexp.MustCompile(`((?:username|password|bearer_token):\s+)(".+"|'.+'|[^\s]+)`)
+	patAuthLine   = regexp.MustCompile(`((?:password|bearer_token|secret_key):\s+)(".+"|'.+'|[^\s]+)`)
```
Comment (fabxc) [inline config/config.go]: Why was username removed here?

Comment (brian-brazil) [inline config/config.go]: Username is not a secret, so it shouldn't be hidden.

Comment (fabxc) [inline config/config.go]: That's a very subjective notion and I cannot see many reasons to have it on the status page. So hiding it is safer and without any real downside.

Comment (brian-brazil) [inline config/config.go]: The downside is that you can't see what username is in use, which makes debugging and operations harder.

Usernames are generally not considered secret, and I'd be very wary of any system that assumes they are secret.

Comment (fabxc) [inline config/config.go]: Fine with me, not strong feelings here.
Inserting potentially impactful changes in one-liners between commits, that are hundreds of lines of unrelated changes, should be avoided though.

Comment (mjibson) [inline config/config.go]: I concur with @fabxc that the removal of username should be made in another PR.

Comment (brian-brazil) [inline config/config.go]: Let's do that then.

File: retrieval/discovery/ec2.go (hunk 14 above)
Comment (mjibson) [inline retrieval/discovery/ec2.go]: I did a test using the existing Source code which uses the Region. I duplicated an `ec2_sd_config` block so there were two identical ones in the same block. The same hosts showed up twice (which is what I expected). I'm thus not convinced we need to change the Source code to use the instance-id. Or I'm misunderstanding something again.

**Comment by mjibson** [comment]
Updated with restored username filter.

File: config/config.go [hunk 18]
```diff
@@ -664,6 +672,34 @@ func (c *KubernetesSDConfig) UnmarshalYAML(unmarshal func(interface{}) error) er
...
+	type plain EC2SDConfig
+	err := unmarshal((*plain)(c))
+	if err != nil {
+		return err
+	}
+	if c.Region == "" {
+		return fmt.Errorf("EC2 SD configuration requires a region")
+	}
+	if c.AccessKey == "" || c.SecretKey == "" {
+		return fmt.Errorf("EC2 SD configuration requires both access and secret keys")
```
Comment (brian-brazil) [inline config/config.go]: Access and secret keys aren't required if you'd getting credentials from the environment.

**Comment by mjibson** [comment]
Updated with removed check for access and secret key.

**Comment by brian-brazil** [comment]
:+1:

**Comment by mjibson** [comment]
Status of this? Waiting until after the 0.16.0 release to merge?

**Comment by fabxc** [comment]
I will review this today. If nothing special comes up it will be in the next release, flagged as beta like the other new SDs.

prometheus/docs#221 has to be completed though.

**Comment by fabxc** [comment]
That are quite a few dependencies, but they are pretty self-contained in general and I think we have a good system in place now so we can ditch the whole "dependencies are bad" ideology once and for all.

File: retrieval/discovery/ec2.go [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+		for _, r := range p.Reservations {
+			for _, inst := range r.Instances {
+				if inst.PrivateIpAddress == nil {
+					continue
+				}
+				labels := model.LabelSet{
+					ec2Label + "instance_id": model.LabelValue(*inst.InstanceId),
+				}
+				if inst.PublicIpAddress != nil {
+					labels[ec2Label+"public_ip"] = model.LabelValue(*inst.PublicIpAddress)
```
Comment (fabxc) [inline retrieval/discovery/ec2.go]: All those labels should be constants defined above.

File: retrieval/discovery/ec2.go (hunk 14 above)
Comment (fabxc) [inline retrieval/discovery/ec2.go]: Seems like the original problem was never tackled here.
We can discuss changes to the interface – but either someone proposes and migrates those changes or this implementation has to fulfill the contract the current interface requires.

File: retrieval/discovery/ec2.go [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+
+	// Get an initial set right away.
+	if err := ed.refresh(ch); err != nil {
+		log.Error(err)
+	}
+
+	for {
+		select {
+		case <-ticker.C:
+			if err := ed.refresh(ch); err != nil {
```
Comment (fabxc) [inline retrieval/discovery/ec2.go]: I recommend not passing down the channel as it is only used for a single send operation. The function can simply return the target group, which keeps the interface simpler.
That also makes testing easier, which I would of course prefer having in place when adding a new SD mechanism.

File: retrieval/discovery/ec2.go [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+)
+
+// EC2Discovery periodically performs EC2-SD requests. It implements
+// the TargetProvider interface.
+type EC2Discovery struct {
+	aws            *aws.Config
+	done           chan struct{}
+	interval       time.Duration
+	port           int
+	instanceFilter []*ec2.Filter
```
Comment (fabxc) [inline retrieval/discovery/ec2.go]: What exactly are those? They are used below but never set. I assume there are plans to make them configurable later – if so, what would that look like?

File: retrieval/discovery/ec2.go [hunk 22]
```diff
@@ -0,0 +1,126 @@
...
+	}
+}
+
+// Sources implements the TargetProvider interface.
+func (ed *EC2Discovery) Sources() []string {
+	return []string{*ed.aws.Region}
+}
+
+func (ed *EC2Discovery) refresh(ch chan<- *config.TargetGroup) error {
+	EC2 := ec2.New(ed.aws)
```
Comment (fabxc) [inline retrieval/discovery/ec2.go]: uppercase naming for local variables is not idiomatic Go.

File: retrieval/discovery/ec2.go (hunk 21 above)
Comment (fabxc) [inline retrieval/discovery/ec2.go]: Okay, I sense that those filters only operate on tags (actually labels) too, which would be logical duplication of our relabeling, so this is safe to be dropped at the cost of initial convenience.

In that case let's drop the field here and make it clear how filtering is supposed to be done for us via a brief comment in the code of `refresh()` and by mention and example in the documentation.

**Comment by fabxc** [comment]
Added comments – the very least the config should be tested. There are plenty of short examples.

A general test of the SD functionality would be helpful too, of course.

File: retrieval/discovery/ec2.go (hunk 21 above)
Comment (mjibson) [inline retrieval/discovery/ec2.go]: This was a thing we decided to remove a few reviews ago and I just forgot to remove the variable.

File: retrieval/discovery/ec2.go (hunk 14 above)
Comment (mjibson) [inline retrieval/discovery/ec2.go]: I thought it was handled. I verified that even if multiple instances are returned via the same config, they would not interfere with each other. My current understanding is this is functioning correctly. If there's something broken I'll need some explanation.

Comment (fabxc) [inline retrieval/discovery/ec2.go]: Oh, sorry – earlier today I somehow saw this as an empty string slice, which it is not.
It is very much correct this way.

As @brian-brazil pointed out, the `TargetProvider` interface is not equally convenient for every SD mechanism.
This is actually one of the cases where it is convenient and simple to use.

**Comment by mjibson** [comment]
Addressed your comments: better var names, const, config tests, no passing of chan, removal of unused properties. Will add docs.

**Comment by fabxc** [comment]
Wonderful, thanks a lot! :+1:
